pkg/server/registry/apigroups/acorn/apps/events.go

```diff
@@ -0,0 +1,198 @@
+package apps
+
+import (
+	"context"
+	"encoding/json"
+
+	apiv1 "github.com/acorn-io/acorn/pkg/apis/api.acorn.io/v1"
+	v1 "github.com/acorn-io/acorn/pkg/apis/internal.acorn.io/v1"
+	"github.com/acorn-io/acorn/pkg/event"
+	"github.com/acorn-io/mink/pkg/strategy"
+	"github.com/acorn-io/mink/pkg/types"
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	AppCreateEventType     = "AppCreate"
+	AppDeleteEventType     = "AppDelete"
+	AppSpecUpdateEventType = "AppSpecUpdate"
+)
+
+// AppSpecCreateEventDetails captures additional info about the creation of an App.
+type AppSpecCreateEventDetails struct {
+	// ResourceVersion is the resourceVersion of the App created.
+	ResourceVersion string `json:"resourceVersion"`
+}
+
+// AppSpecDeleteEventDetails captures additional info about the deletion of an App.
+type AppSpecDeleteEventDetails struct {
+	// ResourceVersion is the resourceVersion of the App deleted.
+	ResourceVersion string `json:"resourceVersion"`
+}
+
+// AppSpecUpdateEventDetails captures additional info about an update to an App Spec.
+type AppSpecUpdateEventDetails struct {
+	// ResourceVersion is the resourceVersion of the updated App.
+	ResourceVersion string `json:"resourceVersion"`
+
+	// OldSpec is the spec of the App before the update.
+	OldSpec v1.AppInstanceSpec `json:"oldSpec"`
+
+	// Patch is a JSON Merge Patch that describes all changes made to OldSpec by the respective update.
+	// See: https://datatracker.ietf.org/doc/html/rfc7386
+	Patch json.RawMessage `json:"patch"`
+}
+
+type eventRecordingStrategy struct {
+	strategy.CompleteStrategy
+	recorder event.Recorder
+}
+
+func newEventRecordingStrategy(s strategy.CompleteStrategy, recorder event.Recorder) *eventRecordingStrategy {
+	return &eventRecordingStrategy{
+		CompleteStrategy: s,
+		recorder:         recorder,
+	}
+}
+
+func (s *eventRecordingStrategy) Create(ctx context.Context, obj types.Object) (types.Object, error) {
+	created, err := s.CompleteStrategy.Create(ctx, obj)
+	if err != nil {
+		// Return created because CompleteStrategy.Create is a black box; i.e. we can't assume
+		// updated is nil when err is non-nil.
```

Comment:
nit (and same below)
```suggestion
		// created is nil when err is non-nil.
```